### ISSUE TITLE
Combine all `with-*` attributes into a single attribute (`ssr-slots`)

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#261]
Currently we have 8 different `with-*` attributes:

1. `with-label` (7 components)
1. `with-help-text` (7 components)
1. `with-header` (3 components)
1. `with-footer` (3 components)
1. `with-image` (`<wa-card>`)
1. `with-submenu` (`<wa-menu-item>`)
1. `with-prefix` (`<wa-radio-button>`)
1. `with-suffix` (`<wa-radio-button>`)

These are only required for SSR (and possibly React?), to communicate which slots are filled. @KonnorRogers could elaborate on the details of why these are needed and why for these specific slots.

I wonder, could we reduce the API surface of this wart and make it more predictable by combining these into a single `with` attribute that takes a space-separated list of slot names? Then we only need to document this once and authors have to only learn this once. The only variable is which slots need this kind of hint, which we could even include in the Slots table. Additionally, it aids readability before it ensures they are all in one place, rather than scattered among other attributes.

Before & after:
```html
<wa-dialog label="Dialog" with-header with-footer class="dialog-overview">
<wa-dialog label="Dialog" with="header footer" class="dialog-overview">
<wa-dialog label="Dialog" has="header footer" class="dialog-overview"> <!-- alt -->
```

**Comment by KonnorRogers** [comment]
> Currently we have 8 different `with-*` attributes:
> 
>     1. `with-label` (7 components)
> 
>     2. `with-help-text` (7 components)
> 
>     3. `with-header` (3 components)
> 
>     4. `with-footer` (3 components)
> 
>     5. `with-image` (`<wa-card>`)
> 
>     6. `with-submenu` (`<wa-menu-item>`)
> 
>     7. `with-prefix` (`<wa-radio-button>`)
> 
>     8. `with-suffix` (`<wa-radio-button>`)
> 
> 
> These are only required for SSR (and possibly React?), to communicate which slots are filled. @KonnorRogers could elaborate on the details of why these are needed and why for these specific slots.
> 
> I wonder, could we reduce the API surface of this wart and make it more predictable by combining these into a single `with` attribute that takes a space-separated list of slot names? Then we only need to document this once and authors have to only learn this once. The only variable is which slots need this kind of hint, which we could even include in the Slots table. Additionally, it aids readability before it ensures they are all in one place, rather than scattered among other attributes.
> 
> Before & after:
> 
> ```
> <wa-dialog label="Dialog" with-header with-footer class="dialog-overview">
> <wa-dialog label="Dialog" with="header footer" class="dialog-overview">
> <wa-dialog label="Dialog" has="header footer" class="dialog-overview"> 
> ```

Would `slotted="header footer"` perhaps make more sense?

`has` I worry about being confused with `:has()`, `with` just feels...vague?

**Comment by LeaVerou** [comment]
> Would `slotted="header footer"` perhaps make more sense?

Yes! Only issue I see with that is that it prevents us from expanding it beyond slots (e.g. detecting`data-toggle-nav` in `<wa-page>`).

**Comment by claviska** [comment]
The `with-` attribute pattern was conceived to support SSR, but we were able to maintain slot detection for non-SSR users. That said, if this is only and will only be used for SSR, would it make sense to tie the name to that?

```html
<wa-dialog ssr-slots="with-header with-footer">
```

Either way, I'm fully onboard with consolidating these.

**Comment by LeaVerou** [comment]
If it's only for SSR +1000 for naming it after that (and then we don't need the `with-`, just `ssr-slots="header footer"`. @KonnorRogers is it needed in any other case?

**Comment by KonnorRogers** [comment]
> If it's only for SSR +1000 for naming it after that (and then we don't need the `with-`, just `ssr-slots="header footer"`. @KonnorRogers is it needed in any other case?

Not that I'm aware of. I only converted the form controls, but all of the form controls end up using the JavaScript slot detection once the component connects, so it only matters for SSR for those. Not 100% on the others.

**Comment by LeaVerou** [comment]
Awesome. I'd argue for prioritizing this, since it's currently very misleading and I worry users will think they need to use these regardless since we're using them in our own examples.

**Comment by claviska** [comment]
This makes a couple things more difficult:

- Types are harder to define per-component as compared to separate boolean attributes
- ^ affects IntelliSense
- We don't have a plan for documenting them as a single property outside of stuffing all the names in a description

I do think there's a way to type them in TypeScript, but it's not obvious to me and none of my AI searches yielded anything comfortable.

Let's reevaluate this after June to see how many users even care about SSR.

**Comment by KonnorRogers** [comment]
> This makes a couple things more difficult:
> 
>     * Types are harder to define per-component as compared to separate boolean attributes
> 
>     * ^ affects IntelliSense
> 
>     * We don't have a plan for documenting them as a single property outside of stuffing all the names in a description
> 
> 
> I do think there's a way to type them in TypeScript, but it's not obvious to me and none of my AI searches yielded anything comfortable.
> 
> Let's reevaluate this after June to see how many users even care about SSR.

There is a way to get TypeScript to recognize it for when we get there. It involves using "Tagged Template Literals" grabbing a union of the "SSR_SLOTS" property on the component. The type itself is kind of annoying to get right, but is doable, and usually you just chuck it in a separate type file. Its doable, and when we get there, feel free to tag me and I can help make it happen. I'm not 100% how it will play with CEM and we may need to use Burton's TypeScript package for CEM, but I wouldn't consider that a major blocker.

> We don't have a plan for documenting them as a single property outside of stuffing all the names in a description

^ this is the harder one and may require special handling in our documentation generation.

> Let's reevaluate this after June to see how many users even care about SSR.

SSR is clearly not our focus right now, so i think its good to table it for now.

**Comment by LeaVerou** [comment]
If we're going to table this for now, I wonder if we could remove all the `with-*` usage from our examples so that folks don't get the impression that they _must_ use these (regardless of SSR). That was my main motivation behind prioritizing this: that the current names are misleading. But we've updated their descriptions to explain they're only needed for SSR, so if we don't use them in our examples I think we can totally defer this.

**Comment by claviska** [comment]
Good call. I can handle that since we’re not using SSR anymore on the website.

**Comment by claviska** [comment]
We're actually repurposing `with-` and adding `without-` as boolean attributes to toggle content, so this is no longer relevant.
